File: packages/elixir/lib/kreuzberg/code_symbol_info.ex
defmodule Kreuzberg.CodeSymbolInfo do
  @moduledoc """
  Symbol definition information.

  ## Fields

    * `:name` - Symbol name
    * `:kind` - Symbol kind (e.g. "variable", "constant")
    * `:type_annotation` - Type annotation if present
    * `:span` - Source span
  """

  @type t :: %__MODULE__{
          name: String.t(),
          kind: String.t(),
          type_annotation: String.t() | nil,
          span: Kreuzberg.CodeSpan.t()
        }

  defstruct name: "",
            kind: "",
            type_annotation: nil,
            span: %Kreuzberg.CodeSpan{}

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    span =
      case Map.get(data, "span", Map.get(data, :span)) do
        %Kreuzberg.CodeSpan{} = s -> s
        map when is_map(map) -> Kreuzberg.CodeSpan.from_map(map)
        _ -> %Kreuzberg.CodeSpan{}
      end

    %__MODULE__{
      name: Map.get(data, "name", Map.get(data, :name, "")),
      kind: Map.get(data, "kind", Map.get(data, :kind, "")),
      type_annotation: Map.get(data, "type_annotation", Map.get(data, :type_annotation)),
      span: span
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = info) do
    %{
      "name" => info.name,
      "kind" => info.kind,
      "type_annotation" => info.type_annotation,
      "span" => Kreuzberg.CodeSpan.to_map(info.span)
    }
  end
end

File: packages/php/src/Types/CodeSymbolInfo.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Symbol definition information.
 *
 * @property-read string $name Symbol name
 * @property-read string $kind Symbol kind (e.g. 'variable', 'constant')
 * @property-read string|null $typeAnnotation Type annotation if present
 * @property-read CodeSpan $span Source span
 */
readonly class CodeSymbolInfo
{
    public function __construct(
        public string $name,
        public string $kind,
        public ?string $typeAnnotation,
        public CodeSpan $span,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $name */
        $name = $data['name'] ?? '';

        /** @var string $kind */
        $kind = $data['kind'] ?? '';

        /** @var string|null $typeAnnotation */
        $typeAnnotation = $data['type_annotation'] ?? null;

        /** @var array<string, mixed> $spanData */
        $spanData = $data['span'] ?? [];

        return new self(
            name: $name,
            kind: $kind,
            typeAnnotation: $typeAnnotation,
            span: CodeSpan::fromArray($spanData),
        );
    }
}
